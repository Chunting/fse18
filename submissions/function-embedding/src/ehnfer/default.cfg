[main]
logging_functions = printk mlog_and_bits mlog_not_bits task_pid_nr
tactics = FunctionContext PrePost
