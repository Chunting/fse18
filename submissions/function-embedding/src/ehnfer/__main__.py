#!/usr/bin/env python3

import os
import sys
import argparse
import ehnfer.commands

def main():
    script_dir = os.path.dirname(os.path.realpath(__file__))

    action, kwargs = parse_args()
    action(**kwargs)


def parse_args():
    def help_and_exit(parser):
        parser.print_help()
        sys.exit()

    parser = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument('command', type=str, nargs='?',
                        help="Action to run. Possible commands:\n"
                             "mine\t\trequires --db, can pass sthresh here\n"
                             "\t\t\tpassing --input will copy labels from input file to report\n"
                             "stripnumbers\trequires --db\n"
                             "killnumbers\trequires --db\n"
                        )
    parser.add_argument('--input', type=str,
                        help='Path to input file.')
    parser.add_argument('--output', type=str,
                        help='Path to output file. None for stdout.')
    parser.add_argument('--db', type=str,
                        help='Path to database.')
    parser.add_argument('--support', type=str,
                        help="Support threshold, defaults to 10", required=True)
    parser.add_argument('--similarity', type=str,
                        help="Similarity threshold, defaults to 0.9", required=True)
    parser.add_argument('--model', type=str, help="Path to model file", required=True)

    args = parser.parse_args()

    # Default options
    arg_command = None

    if args.command:
        arg_command = args.command
    else:
        help_and_exit(parser)

    action_kwargs = dict()
    if arg_command == "mine":
        if not args.db:
            help_and_exit(parser)

        action = ehnfer.commands.mine
        action_kwargs["db_file"] = args.db
        action_kwargs["support_threshold"] = int(args.support)
        action_kwargs["similarity_threshold"] = float(args.similarity)
        action_kwargs["model_file"] = args.model
    else:
        help_and_exit(parser)

    return action, action_kwargs

if (__name__ == "__main__"):
    main()
