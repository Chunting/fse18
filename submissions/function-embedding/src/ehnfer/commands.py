from __future__ import print_function
import sys
import os
import itertools
import logging
from collections import defaultdict
import math
import scipy.misc

from ehnfer.handler_db import HandlerDb, Item, ItemType, TacticType, Spec
from gensim.models.keyedvectors import KeyedVectors

def mine(db_file=None, support_threshold=3, similarity_threshold=0.9, model_file=None, num_epochs=10, min_items=2, max_items=3):
    blacklist = set([
        "strcpy",
        "seq_printf",
        "sprintf",
        "__mlog_printk",
        "strcat",
        "strlcpy",
        "init_timer_key",
        "IS_ERR",
        "GFS2_I",
        "OCFS2_I",
        "BTRFS_I",
        "INODE_CACHE",
        "GFS2_SB",
        "EXT4_SB",
        "btrfs_sb"
        ])

    graylist = set([
        "dev_err",
        "dquot_initialize",
        "ocfs2_inode_lock_full_nested",
        "warn_slowpath_null",
        "ocfs2_init_dinode_extent_tree",
        "fs_path_alloc",
        "btrfs_next_leaf",
        "ocfs2_lock_refcount_tree",
        "ocfs2_check_dir_for_entry",
        "ocfs2_prepare_dir_for_insert",
        "warn_slowpath_fmt",
        "ocfs2_init_dealloc_ctxt"
    ])

    # COMMENT out union to see blacklist only
    blacklist = blacklist.union(graylist)

    # Open the handler database and read handler information
    hdb = HandlerDb(db_file)

    logging.basicConfig(format='%(asctime)s %(message)s', datefmt='%m/%d/%Y %I:%M:%S %p', level=logging.INFO)

    # A map from an item to all of the handlers that have that item in their context
    import tempfile
    import subprocess

    # Mine three item closed sets using eclat
    tmp_name_tx = "%s/%s" % (tempfile.gettempdir(), next(tempfile._get_candidate_names()))
    tmp_name_mined = "%s/%s" % (tempfile.gettempdir(), next(tempfile._get_candidate_names()))

    # Call miner
    set_counts = defaultdict(set)

    # Write out sentences to tmp file
    # For each handler
    with open(tmp_name_tx, 'w') as f:
        for h in hdb.handlers.keys():
            # For each pair of items, one from context and one from response
            handler = hdb.handlers[h]

            # Print out handler sentence
            if len(handler.context) == 0 or len(handler.response) == 0:
                continue
            sentence = ["PRE|" + x.name for x in handler.context]
            sentence += (["POST|" + x.name for x in handler.response])
            f.write(" ".join(sentence))
            f.write("\n")
    print("SENTENCES AT:", tmp_name_tx)

    # Run eclat
    support_param = "-s-%s" % support_threshold
    max_item_param = "-n{}".format(max_items)
    min_item_param = "-m{}".format(min_items)
    script_dir = os.path.dirname(os.path.realpath(__file__))
    subprocess.call(
        ['/program2vec/lib/eclat/eclat/src/eclat', '-ts', support_param, max_item_param, min_item_param, '-v %a', tmp_name_tx, tmp_name_mined])
#    os.remove(tmp_name_tx)

    # Read in the frequent itemsets
    rules = set()
    with open(tmp_name_mined, "r") as f:
        raw_mined = f.readlines()
    os.remove(tmp_name_mined)
    raw_mined = [x.rstrip() for x in raw_mined]
    for t in raw_mined:
        items = t.split()
        rule_support = int(items.pop())

        context_functions = set()
        response_functions = set()
        for item in items:
            prefix, item_name = item.split("|")
            if item_name in blacklist:
                continue
            if prefix == "PRE":
                context_functions.add(item_name)
            else:
                response_functions.add(item_name)
        if len(context_functions) != 0 and len(response_functions) != 0:
            rules.add(AssociationRule(context=context_functions, response=response_functions, support=rule_support))

    # Convert to specs, but support is already done

    specs = [Spec(set([x])) for x in rules]
    print(support_threshold, scipy.misc.comb(len(specs), 2))

    print(len(specs))
    single_item_responses = defaultdict(set)
    for s in specs:
        for r in s.rules:
            if len(r.response) == 1:
                item = next(iter(r.response))
                single_item_responses[item].add(s)

    model = KeyedVectors.load_word2vec_format(model_file, binary=False)
    print(len(single_item_responses.keys()))

    # INSIGHT: We only need the raw handler list for merges, so calc. for rules on demand

    spec_sims = []
    for response1, response2 in itertools.combinations(single_item_responses, 2):
        specs1 = single_item_responses[response1]
        specs2 = single_item_responses[response2]

        for spec1, spec2 in itertools.product(specs1, specs2):
            spec_sims.append((spec1, spec2, spec1.similarity(spec2, model)))
    spec_sims.sort(key=lambda x: x[2], reverse=True)

    # Go through the spec pairs with sim above threshold
    # Merge them in order
    total_merges = 0
    for i in xrange(num_epochs):
        specs = set(specs)
        print("Starting epoch {} of {}".format(i+1, num_epochs))
        merged_this_epoch = 0
        done = 0

        new_sims = []
        for ss in spec_sims:
            spec1 = ss[0]
            spec2 = ss[1]
            sim = ss[2]
            print(sim)

            done += 1
            print(done, len(spec_sims))
            if sim >= similarity_threshold:
                if spec1 not in specs or spec2 not in specs:
                    continue

                # Add merged spec to list of all specs, and remove individual specs
                merged = spec1.merge(spec2, hdb)

                for s in specs:
                    new_sims.append((merged, s, merged.similarity(s, model)))

                specs.add(merged)
                specs.remove(spec1)
                specs.remove(spec2)

                # Recompute support for merged specifications
                merged_this_epoch += 1
            else:
                break

        spec_sims += new_sims
        spec_sims.sort(key=lambda x: x[2], reverse=True)
        specs = list(specs)
        specs.sort(key=lambda x: x.support, reverse=True)

        total_merges += merged_this_epoch
        print("{} merges".format(merged_this_epoch))
        if merged_this_epoch == 0:
            break

    specs.sort(key=lambda x: x.support, reverse=True)

    return specs

class Spec:
    def __init__(self, rules):
        """
        :param rules: A set of AssociationRules
        """
        self.rules = rules
        self.confidence = None

    def min_similarity(self, other, model):
        min_sim = 1.0
        for r1, r2 in itertools.product(self.rules, other.rules):
            sim = r1.similarity(r2, model)
            if sim < min_sim:
                min_sim = sim
        return min_sim

    def similarity(self, other, model):
        return self.min_similarity(other, model)

    def merge(self, other, hdb):
        new_rules = self.rules.union(other.rules)

        for rule in new_rules:
            for h in hdb.handlers.keys():
                handler = hdb.handlers[h]
                handler_context = set([x.name for x in handler.context])
                handler_response = set([x.name for x in handler.response])

                if rule.context.issubset(handler_context) and rule.response.issubset(handler_response):
                    rule.supporting_handlers.add(handler)

        # For now just take the union
        # Let's see what happens
        return Spec(new_rules)

    def __repr__(self):
        ret = "support: {}\n".format(self.support)
        ret += "\n".join([str(r) for r in self.rules])
        return ret

    @property
    def support(self):
        support = 0
        supporting_handlers = set()

        if len(self.rules) == 1:
            return next(iter(self.rules)).support

        for r in self.rules:
            # Merged specs must have supporting handlers populated for rules
            if len(self.rules) > 1:
                assert len(r.supporting_handlers) > 0
            for h in r.supporting_handlers:
                supporting_handlers.add(h)
        return len(supporting_handlers)

    @property
    def id(self):
        return id(self)


class AssociationRule:
    def __init__(self, context, response, support):
        self.context = context
        self.response = response
        self.__support = support
        self.judgement = None

        self.supporting_handlers = set()
        self.violating_handlers = set()

        self.rank = None

    @property
    def support(self):
        if len(self.supporting_handlers) == 0:
            return self.__support
        else:
            return len(self.supporting_handlers)

    def __repr__(self):
        return "{} -> {} s{}".format(self.context, self.response, self.support)

    def similarity(self, other, model):
        try:
            if self.context == other.context and len(self.response) == 1:
                return model.similarity(next(iter(self.response)), next(iter(other.response)))
            elif self.response == other.response and len(self.context) == 1:
                return model.similarity(next(iter(self.context)), next(iter(other.context)))
            else:
                return 0.0
        except:
            # Not in model
            return 0.0

    @property
    def id(self):
        return id(self)

