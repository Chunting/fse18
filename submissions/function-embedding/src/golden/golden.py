import base64
import os
import StringIO
from io import BytesIO
from collections import namedtuple

import boto3
import flask
from gensim.models import KeyedVectors
import itertools
import matplotlib.pyplot as plt

import walker.metric
import walker.golden
import walker.data
import walker.visualization

import ehnfer.commands

app = flask.Flask(__name__)
app.config.from_object(__name__)

params = 'arxiv_vmlinux_length100_walks100_interproc1_window1_mincount5_dimensions300'
interprocedural_model_file = '/data/current.model'

# Local copy of s3://program2vec/fse/flat.vec
flat_model_file = '/data/flat.vec'
intraprocedural_model_file = '/home/daniel/tmp/intra.vec'

# Local copy of s3://program2vec/fse/bias2-0_removecross1_length100_numwalks100_interproc1_window1_mincount5_dim300.model
biased_model_file = '/home/daniel/tmp/bias2-remove1.vec'
#local_clusters_file = '/home/daniel/tmp/clusters.txt'
local_clusters_file = '/home/daniel/tmp/flat3500.clusters'

sound_handler_db = '/data/pcisound.sqlite'
fs_handler_db = '/home/daniel/tmp/5fs-unmerged.sqlite'

handler_db=sound_handler_db
lower_support_threshold = 3
similarity_threshold = 0.6

local_marked_file = '/data/marked_specs.txt'

GOLDEN_INPUT_ALL = walker.golden.GoldenInput(
    use_underscore=True,
    use_bootstrap=True,
    clusters_file='arxiv/clusters/{}_k3500.txt'.format(params),
    reviewed_files = ['aditya-relations.txt',
                      'cindy-relations.txt',
                      'aditya-dd-relations.txt',
                      'two-function-relations.txt',
                      'golden-set-review-or-something-relations.txt',
                       'sys-relations.txt',
                      '2018-02-19-relations.txt',
                      '2018-02-19-b-relations.txt',
                      'daniel-relations.txt',
    ]
)

GOLDEN_INPUT_NOBOOT = walker.golden.GoldenInput(
    use_underscore=True,
    use_bootstrap=False,
    clusters_file='arxiv/clusters/{}_k3500.txt'.format(params),
    reviewed_files = ['aditya-relations.txt',
                      'cindy-relations.txt',
                      'aditya-dd-relations.txt',
                      'two-function-relations.txt',
                      'golden-set-review-or-something-relations.txt',
                       'sys-relations.txt',
                      '2018-02-19-relations.txt',
                      '2018-02-19-b-relations.txt',
                      'daniel-relations.txt'
    ]
)

GOLDEN_INPUT_UNDERSCORES = walker.golden.GoldenInput(
    use_underscore=True,
    use_bootstrap=False,
    clusters_file='arxiv/clusters/{}_k3500.txt'.format(params),
    reviewed_files = [
       # 'aditya-relations.txt',
       #               'cindy-relations.txt',
       #               'aditya-dd-relations.txt',
       #               'two-function-relations.txt',
       #               'golden-set-review-or-something-relations.txt',
       #                'sys-relations.txt',
       #               '2018-02-19-relations.txt',
       #               '2018-02-19-b-relations.txt',
       #               'daniel-relations.txt'
    ]
)

GOLDEN_INPUT_MANUAL = walker.golden.GoldenInput(
    use_underscore=False,
    use_bootstrap=False,
    clusters_file='arxiv/clusters/{}_k3500.txt'.format(params),
    reviewed_files = [
        'aditya-relations.txt',
        'cindy-relations.txt',
        'aditya-dd-relations.txt',
        'two-function-relations.txt',
        'golden-set-review-or-something-relations.txt',
        'sys-relations.txt',
        '2018-02-19-relations.txt',
        '2018-02-19-b-relations.txt',
        'daniel-relations.txt'
    ]
)

FSCORE_GOLDEN=GOLDEN_INPUT_UNDERSCORES

@app.route('/')
def home():
    return flask.render_template('home.html')

@app.route('/names')
def names():
    # Hardcode S3 path to defined functions, for now.
    s3 = boto3.resource('s3')
    defined_functions_file = StringIO.StringIO()
    s3.Object('program2vec', 'inputs/definedfunctions.txt').download_fileobj(defined_functions_file)

    defined_functions_file.seek(0)
    defined_functions = walker.data.read_functions(defined_functions_file)

    ### DUPLICATION ####
    # This is duplicated from walker.commands.golden_names
    prefix_groups, suffix_groups = walker.golden.group_by_suffix(defined_functions)

    # Get a list of the most common function name prefixes for each file
    # keywords is a map from keyword to the list of files in which that is a top 3 keyword
    keywords = walker.golden.common_keywords(defined_functions)

    # Function must start with the prefix and be a keyword that is paired with the correct file
    # Threshold is the minimum number of function names that use the keyword
    keyword_filtered_suffix_groups = walker.golden.filter_groups_by_keyword(suffix_groups, keywords, [], threshold=10)

    # Filter out functions that have differing first directory paths
    path_filtered_suffix_groups = walker.golden.filter_groups_by_path(keyword_filtered_suffix_groups)

    underscore_wrappers = walker.golden.underscore_wrappers(defined_functions, double=True, single=True)

    ### END DUPLICATION ####

    output = []
    for g in path_filtered_suffix_groups:
        functions = [f.name for f in g.functions]
        output.append(" ".join(functions))

    for pair in underscore_wrappers:
        output.append(" ".join(pair))

    return flask.render_template('names.html', groups=output)

@app.route('/clusters')
def clusters():
    s3 = boto3.resource('s3')
    clusters_file = StringIO.StringIO()
    s3.Object('program2vec', GOLDEN_INPUT_MANUAL.clusters_file).download_fileobj(clusters_file)

    clusters_file.seek(0)
    clusters_lines = clusters_file.readlines()

    return flask.render_template('clusters.html',
                                 groups=clusters_lines
    )

@app.route('/roc')
def roc():
    figfile = BytesIO()

    golden_output_all = walker.golden.golden(GOLDEN_INPUT_ALL)
    golden_output_noboot = walker.golden.golden(GOLDEN_INPUT_NOBOOT)
    golden_output_underscores = walker.golden.golden(GOLDEN_INPUT_UNDERSCORES)
    golden_output_manual = walker.golden.golden(GOLDEN_INPUT_MANUAL)

    num_true = None
    num_false = None

    flat_model = KeyedVectors.load_word2vec_format(flat_model_file, binary=False)
    #interproc_model = KeyedVectors.load_word2vec_format(interprocedural_model_file, binary=False)
    #intra_model = KeyedVectors.load_word2vec_format(intraprocedural_model_file, binary=False)
    biased_model = KeyedVectors.load_word2vec_format(biased_model_file, binary=False)

    #all_scores, sims, true_done, false_done = walker.metric.multi_roc([interproc_model, intra_model, flat_model],
    #                                              golden_output_all,
    #                                              ["interprocedural", "intraprocedural", "flat"],
    #                                              num_true,
    #                                              num_false)

    #print(true_done, false_done)
    #print(len(sims))

    plt.savefig(figfile, format='png')
    plt.clf()
    figfile.seek(0)
    all_figdata_png = base64.b64encode(figfile.getvalue())

    #noboot_scores, sims, true_done, false_done = walker.metric.multi_roc([interproc_model, intra_model, flat_model],
    #                                              golden_output_noboot,
    #                                              ["interprocedural", "intraprocedural", "flat"],
    #                                              num_true,
    #                                              num_false)

    #print(true_done, false_done)
    #print(len(sims))

    plt.savefig(figfile, format='png')
    plt.clf()
    figfile.seek(0)
    noboot_figdata_png = base64.b64encode(figfile.getvalue())

    underscore_scores, sims, true_done, false_done, unique_functions = walker.metric.multi_roc([biased_model, flat_model],
                                                  golden_output_underscores,
                                                  ["path-based", "flat"],
                                                  num_true,
                                                  num_false)

    print(true_done, false_done, unique_functions)

    plt.savefig(figfile, format='png')
    plt.clf()
    figfile.seek(0)
    underscore_figdata_png = base64.b64encode(figfile.getvalue())


    manual_scores, sims, true_done, false_done, unique_functions = walker.metric.multi_roc([biased_model, flat_model],
                                                  golden_output_manual,
                                                  ["path-based", "flat"],
                                                  num_true,
                                                  num_false)

    print(true_done, false_done, unique_functions)

    plt.savefig(figfile, format='png')
    plt.clf()
    figfile.seek(0)
    manual_figdata_png = base64.b64encode(figfile.getvalue())

    return flask.render_template('roc.html',
                                 all_plot=None,
                                 noboot_plot=None,
                                 underscore_plot=underscore_figdata_png,
                                 manual_plot=manual_figdata_png,
                                 all_scores=[],
                                 noboot_scores=[],
                                 underscore_scores=underscore_scores,
                                 manual_scores=manual_scores,
                                 similarities=[],
    )

@app.route('/golden')
def golden():
    # These should be actual inputs

    golden_output = walker.golden.golden(GOLDEN_INPUT_MANUAL)
    must_equiv_classes = golden_output.must_equiv_classes

    output = StringIO.StringIO()
    largest_class = 0
    for c in must_equiv_classes:
        if len(c) > largest_class:
            largest_class = len(c)
        output.write("{}<br><br>\n".format(" ".join([x.name for x in c])))
    output.seek(0)

    num_classes = len(must_equiv_classes)
    print("Largest class,", largest_class)

    return flask.render_template('golden.html',
                                 groups=output,
                                 params=params,
                                 true_relations=golden_output.num_true_relations,
                                 false_relations=golden_output.num_false_relations,
                                 num_classes=num_classes,
                                 use_underscore_names=GOLDEN_INPUT_ALL.use_underscore,
                                 use_bootstrap=GOLDEN_INPUT_ALL.use_bootstrap
    )

@app.route('/specs', methods=['GET'])
def specs():
    specs = ehnfer.commands.mine(db_file=handler_db,
                        support_threshold=lower_support_threshold,
                        similarity_threshold=similarity_threshold,
                        model_file=interprocedural_model_file,
                        num_epochs=10)

    # Read in the marked file
    judgements = dict()
    with open(local_marked_file, 'r') as f:
        lines= f.readlines()
    for l in lines:
        judgement, context, response = l.strip().split()
        if judgement == "T":
            judgement = True
        elif judgement == "?":
            judgement = None
        else:
            judgement = False
        judgements[(context, response)] = judgement

    # Commented sorts by support
    # Uncommented sorts by confidence
    #specs.sort(key=lambda x: x.confidence, reverse=True)

    rank = 1
    for s in specs:
        for r in s.rules:
            r.rank = rank
            rank += 1

#    sorted_judgements = []
#    for spec in specs:
#        # Calculate average precision of entire set, leaving gaps
#        for rule in spec.rules:
#            if (rule.item1.name, rule.item2.name) in judgements:
#                j = judgements[(rule.item1.name, rule.item2.name)]
#                rule.judgement = j
#                sorted_judgements.append(j)

    # Parse pairs, keeping the last in the file
    #if 'csv' in flask.request.args:
    #    return _specs_csv(specs)

#    avgprecision = walker.metric.avgprecision(sorted_judgements)

    return flask.render_template('specs.html', specs=specs, avgprecision=0.0)


def _specs_csv(specs):
    import csv

    si = StringIO.StringIO()
    cw = csv.writer(si)

    for s in specs:
        for r in s.rules:
            cw.writerow([r.item1, r.item2, s.support, r.support])

    output = flask.make_response(si.getvalue())
    output.headers["Content-Disposition"] = "attachment; filename=export.csv"
    output.headers["Content-type"] = "text/csv"

    return output

@app.route('/specs', methods=['POST'])
def mark_spec():
    # Parse the button ID
    button_id = flask.request.form['button']
    judgement, context, response = button_id.split('|')

    # Open the marked file for append and write out the pair
    f = open(local_marked_file, 'a')
    f.write("{} {} {}\n".format(judgement, context, response))
    f.close()

    return "Foo"


@app.route('/new')
def new_metric():
    #model = KeyedVectors.load_word2vec_format(local_model_file, binary=False)

    # Load the clusters
    with open(local_clusters_file, 'r') as f:
        clusters = [set(x.strip().split()) for x in f.readlines()]

    metric = walker.golden.new_metric(GOLDEN_INPUT, clusters)

    precision = float(metric.true_positives) / (metric.true_positives + metric.false_positives)
    recall = float(metric.true_positives) / (metric.true_positives + metric.false_negatives)
    f1 = 2 * precision * recall / (precision + recall)

    return flask.render_template('new.html',
                                 metric=metric,
                                 precision=precision,
                                 recall=recall,
                                 f1=f1,
    )

# TODO!: Print list of golden cluster functions that are not in the model (filtered out)
@app.route('/fscore')
def fscore():
    # Load the clusters
    with open(local_clusters_file, 'r') as f:
        clusters = [set(x.strip().split()) for x in f.readlines()]

    golden_output = walker.golden.golden(FSCORE_GOLDEN)
    must_equiv_classes = golden_output.must_equiv_classes

    # function_classes is just a reformatting of golden set into dict
    function_classes = dict()
    golden_set_size = 0
    for i in xrange(len(must_equiv_classes)):
        must_equiv_class = must_equiv_classes[i]
        for f in must_equiv_class:
            golden_set_size += 1
            function_classes[f.name] = i

    # Create a results class...lol
    results = walker.data.purity(function_classes, clusters)

    f_score = results[6][0][0]
    goldens = results[7]
    avg_precision = results[8][0][0]
    avg_recall = results[9][0][0]

    num_classes = len(goldens)

    # Go through each function in each golden
    # Each function will be a named tuple with the function name and a color to apply
    # Make two lists: functions only in beginning, functions that are not only in the beginning

    return flask.render_template('show_entries.html',
                                 groups=goldens,
                                 f_score=f_score,
#                                 total_golden_size=golden_set_size,
                                 avg_precision=avg_precision,
                                 avg_recall=avg_recall,
                                 params=params,
                                 true_relations=golden_output.num_true_relations,
                                 false_relations=golden_output.num_false_relations,
                                 num_classes=num_classes,
                                 use_underscore_names=GOLDEN_INPUT_MANUAL.use_underscore,
                                 use_bootstrap=GOLDEN_INPUT_MANUAL.use_bootstrap
    )


if __name__ == "__main__":
    app.run(host='0.0.0.0')
