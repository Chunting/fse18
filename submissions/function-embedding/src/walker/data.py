import numpy as np
import pandas as pd

from walker.metric import _labeled_and_clustered
from walker.metric import _precision
import walker.golden
from collections import namedtuple

from walker.metric import _recall
from walker.metric import _F
from walker.metric import _Fscore
from walker.metric import _avg_precision
from walker.metric import _avg_recall

Golden = namedtuple('Golden', 'functions f cluster precision recall extra missing')
Split = namedtuple('Split', 'prefix suffix')


# We do not force a 1:1 mapping because our golden set is a subset of functions clustered
# We are evaluating our performance on the golden set, not each cluster.
# Consider the case of C1=ABCDEF, C2=XYZ, T1=ABC, T2=DEF.
# Map T1 to C1, T2 to C2? That's not fair and does not get at what our golden set is trying to do.
def purity(classes, clusters):
    """
    Full precision / recall / purity numbers
    :param classes: A dictionary of all functions and their reference class IDs {function: class},
    :param clusters: A list of sets
    """

    # IMPORTANT: reference_sets is filtered, label_sets is not.
    #filtered_reference, golden_standard, filtered_clusters = _labeled_and_clustered(classes, clusters)
    golden_standard, _, filtered_clusters = _labeled_and_clustered(classes, clusters)

    golden_standard = [x for x in golden_standard if len(x) > 1]

    num_clusters = len(filtered_clusters)
    num_ref = len(golden_standard)

    per_class_precision = np.zeros((num_ref, num_clusters))
    per_class_recall = np.zeros((num_ref, num_clusters))
    per_class_F = np.zeros((num_ref, num_clusters))

    for r in range(0, num_ref):
        for c in range(0, num_clusters):
            per_class_recall[r][c] = _recall(filtered_clusters[c], golden_standard[r])
            per_class_precision[r][c] = _precision(filtered_clusters[c], golden_standard[r])
            per_class_F[r][c] = _F(filtered_clusters[c], golden_standard[r])

    precision_df = pd.DataFrame(per_class_precision)
    recall_df = pd.DataFrame(per_class_recall)
    f_df = pd.DataFrame(per_class_F)

    # DataFrames are indexed COLUMN, ROW

    # Row index -> Matched cluster
    golden_to_cluster = dict()
    for r in range(num_ref):
        max_in_row = 0.0
        for c in range(num_clusters):
            score = f_df.iloc[r, c]
            if score >= max_in_row:
                max_in_row = score
                golden_to_cluster[r] = c

    max_f = f_df.apply(max, axis=1)

    N = len(reduce(lambda x, y: x.union(y), golden_standard))
    f_score = pd.DataFrame([_Fscore(golden_standard, filtered_clusters, N)])
    avg_precision = pd.DataFrame([_avg_precision(golden_standard, filtered_clusters, N)])
    avg_recall = pd.DataFrame([_avg_recall(golden_standard, filtered_clusters, N)])

    sorted_goldens = []
    for i in range(len(max_f)):
        cluster = filtered_clusters[golden_to_cluster[i]]
        category = golden_standard[i]
        missing = set(category).difference(cluster)
        extra = set(cluster).difference(category)
        sorted_goldens.append(Golden(f=max_f[i], functions=golden_standard[i],
                                     cluster=cluster,
                                     precision=_precision(cluster, category),
                                     recall=_recall(cluster, category),
                                     missing=missing,
                                     extra=extra
                                     )
        )
    sorted_goldens.sort(key=lambda g: g.f)
    #sorted_goldens.sort(key=lambda g: (1.0 - g.f) * len(g.functions), reverse=True)

    return (precision_df, recall_df, f_df, golden_standard, filtered_clusters, max_f, f_score, sorted_goldens,
            avg_precision, avg_recall)

def read_functions(defined_functions_file):
    """
    :param defined_functions_file_path: path to output from definedfunctions LLVM pass
    :return: list of namedtuples
    """
    lines = [x.strip() for x in defined_functions_file.readlines()]
    fns_paths = [(x.split()[0], x.split()[1]) for x in lines]

    functions = []
    for path, name in fns_paths:
        functions.append(walker.golden.Function(name=name, path=path))
    return functions

def split_function_with_keyword(function_name, keyword):
   """
   Splits a function name using a specific keyword. The split is performed at the end of the keyword,
   yielding a prefix / suffix split of (*_)keyword_suffix

   :param function_name: string name of function
   :return: None or a Split namedtuple
   """
   num_underscores = function_name.count('_')
   if num_underscores == 0:
       return None

   name_parts = function_name.split("_")
   try:
       idx = name_parts.index(keyword)
   except ValueError:
       return None

   prefix = "_".join(name_parts[:idx+1])
   suffix = "_".join(name_parts[idx+1:])

   return Split(prefix, suffix)
