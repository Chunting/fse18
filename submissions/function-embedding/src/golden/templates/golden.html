{% extends "layout.html" %}

{% block body %}

<div class=page>
<h1>Computed Golden Set (sum of true relations)</h1>

Parameters: {{ params }}<br><br> 

Using underscore names: {{ use_underscore_names }}<br><br>
Using bootstrap: {{ use_bootstrap }}<br><br>

Total golden set size: <br>
<!-- {{ total_golden_size }} functions<br> -->
{{ true_relations }} must relations<br>
{{ false_relations }} must not relations<br>
{{ num_classes }} equivalence classes<br>
<br>

<hr>

<ul class="groups">
    {% for group in groups %}
    <li>{{ group |safe }}
    {% else %}
    <li><em>No groups to display.</em> 
    {% endfor %}
</ul>
{% endblock %}